src/precompiles/native.rs

```diff
@@ -144,91 +190,138 @@ impl ExitToEthereum {
-    fn run(input: &[u8], target_gas: u64, _context: &Context) -> PrecompileResult {
+    fn run(
+        input: &[u8],
+        target_gas: u64,
+        context: &Context,
+        _state: &mut S,
+        _is_static: bool,
+    ) -> PrecompileResult {
         if Self::required_gas(input)? > target_gas {
             return Err(ExitError::OutOfGas);
         }
 
-        Ok((ExitSucceed::Returned, Vec::new(), 0))
+        Ok(PrecompileOutput {
+            exit_status: ExitSucceed::Returned,
+            output: Vec::new(),
+            cost: 0,
+            logs: Vec::new(),
+        })
     }
 
     #[cfg(feature = "exit-precompiles")]
-    fn run(input: &[u8], target_gas: u64, context: &Context) -> PrecompileResult {
+    fn run(
+        input: &[u8],
+        target_gas: u64,
+        context: &Context,
+        state: &mut S,
+        _is_static: bool,
+    ) -> PrecompileResult {
         if Self::required_gas(input)? > target_gas {
             return Err(ExitError::OutOfGas);
         }
 
-        let (nep141_address, serialized_args) = if context.apparent_value != U256::from(0) {
-            // ETH transfer
-            //
-            // Input slice format:
-            //      eth_recipient (20 bytes) - the address of recipient which will receive ETH on Ethereum
-
-            let eth_recipient: String = hex::encode(input);
-
-            if eth_recipient.len() == 20 {
-                (
-                    crate::sdk::current_account_id(),
-                    crate::prelude::format!(
-                        r#"{{"amount": "{}", "recipient": "{}"}}"#,
-                        context.apparent_value.as_u128(),
-                        eth_recipient
-                    ),
-                )
-            } else {
-                return Err(ExitError::Other(Cow::from("ERR_INVALID_RECIPIENT_ADDRESS")));
+        // TODO(MarX): After calling directly function withdraw in Tester.sol
+        //   This function is called with is_static = true
+        //   Figure out if this needs to be fixed in EVM, or the way
+        //   that is being used to determine if a function is called in static
+        //   mode is incorrect.
+        // if is_static {
+        //     return Err(ExitError::Other(Cow::from("ERR_INVALID_IN_STATIC")));
+        // }
+
+        // First byte of the input is a flag, selecting the behavior to be triggered:
+        //      0x0 -> Eth transfer
+        //      0x1 -> Erc20 transfer
+        let mut input = input;
+        let flag = input[0];
+        input = &input[1..];
+
+        let (nep141_address, serialized_args) = match flag {
+            0x0 => {
+                // ETH transfer
+                //
+                // Input slice format:
+                //      eth_recipient (20 bytes) - the address of recipient which will receive ETH on Ethereum
+
+                let eth_recipient: String = hex::encode(input);
+
+                if eth_recipient.len() == 20 {
+                    (
```

Comment:
The `input` is 20 bytes long, but the `eth_recipient` is 40 bytes long, as this is a hex-encoded address.

```suggestion
                if input.len() == 20 {
                    let eth_recipient: String = hex::encode(input);
                    (
```

Reply:
Good catch, this is why we need end 2 end test before merging! I'm finishing them.